daemons/fenced/fenced_remote.c

```diff
@@ -2030,19 +2042,257 @@ process_remote_stonith_exec(xmlNode * msg)
+ * \param[in] target            Optionally limit to certain fence-target
+ *
+ * \return The fence-history as xml
+ */
+static xmlNode *
+stonith_local_history_diff(GHashTable *remote_history,
+                           gboolean add_id,
+                           const char *target)
+{
+    xmlNode *history = NULL;
+    int cnt = 0;
+
+    if (remote_op_list) {
+            GHashTableIter iter;
+            remote_fencing_op_t *op = NULL;
+
+            history = create_xml_node(NULL, F_STONITH_HISTORY_LIST);
+
+            g_hash_table_iter_init(&iter, remote_op_list);
+            while (g_hash_table_iter_next(&iter, NULL, (void **)&op)) {
+                xmlNode *entry = NULL;
+
+                if (remote_history &&
+                    g_hash_table_lookup(remote_history, op->id)) {
+                    continue; /* skip entries broadcasted already */
+                }
+
+                if (target && strcmp(op->target, target) != 0) {
+                    continue;
+                }
+
+                cnt++;
+                crm_trace("Attaching op %s", op->id);
+                entry = create_xml_node(history, STONITH_OP_EXEC);
+                if (add_id) {
+                    crm_xml_add(entry, F_STONITH_REMOTE_OP_ID, op->id);
+                }
+                crm_xml_add(entry, F_STONITH_TARGET, op->target);
+                crm_xml_add(entry, F_STONITH_ACTION, op->action);
+                crm_xml_add(entry, F_STONITH_ORIGIN, op->originator);
+                crm_xml_add(entry, F_STONITH_DELEGATE, op->delegate);
+                crm_xml_add(entry, F_STONITH_CLIENTNAME, op->client_name);
+                crm_xml_add_int(entry, F_STONITH_DATE, op->completed);
+                crm_xml_add_int(entry, F_STONITH_STATE, op->state);
+            }
+    }
+
+    if (cnt == 0) {
+        free_xml(history);
+        return NULL;
+    } else {
+        return history;
+    }
+}
+
+/*!
+ * \internal
+ * \brief Merge fence-history coming from remote into local history
+ *
+ * \param[in] history   Hash-table holding remote history to be merged in
+ */
+static void
+stonith_merge_in_history_list(GHashTable *history)
+{
+    GHashTableIter iter;
+    remote_fencing_op_t *op = NULL;
+    gboolean updated = FALSE;
+
+    if (!history) {
+        return;
+    }
+
+    init_remote_op_list();
+
+    g_hash_table_iter_init(&iter, history);
+    while (g_hash_table_iter_next(&iter, NULL, (void **)&op)) {
+        remote_fencing_op_t *stored_op =
+            g_hash_table_lookup(remote_op_list, op->id);
+
+        if (stored_op) {
+            continue; /* skip over existant - state-merging migh be desirable */
+        }
+
+        updated = TRUE;
+        g_hash_table_iter_steal(&iter);
+        CRM_LOG_ASSERT(g_hash_table_insert(remote_op_list, op->id, op));
+    }
+    if (updated) {
+        do_stonith_notify(0, T_STONITH_NOTIFY_HISTORY, 0, NULL);
+    }
+    g_hash_table_destroy(history); /* remove what is left */
+}
```

Comment:
I think a separate fenced_history.c with relevant code is justified now

Reply:
Yep, seems to be fairly independent from the rest of the code in fenced_remote.c.
On the first sight only thing we'd have to make global somehow is remote_op_list, free_remote_op and init_remote_op_list. (e.g. stonith_remote_op_list, stonith_init_remote_op_hash_table(GHashTable **))